aptos-move/framework/aptos-framework/sources/Table.move

```diff
@@ -1,152 +1,108 @@
-/// This module provides a temporary solution for tables by providing a layer on top of Vector
+/// Type of large-scale storage tables.
 module AptosFramework::Table {
     use Std::Errors;
-    use Std::Option;
-    use Std::Vector;
 
-    const EKEY_ALREADY_EXISTS: u64 = 0;
-    const EKEY_NOT_FOUND: u64 = 1;
+    // native code raises this with Errors::invalid_arguments()
+    const EALREADY_EXISTS: u64 = 100;
+    // native code raises this with Errors::invalid_arguments()
+    const ENOT_FOUND: u64 = 101;
+    const ENOT_EMPTY: u64 = 102;
 
-    struct Table<Key: store, Value: store> has store {
-      data: vector<TableElement<Key, Value>>,
+    /// Type of tables
+    struct Table<phantom K, phantom V> has store {
+        handle: u128,
+        length: u64,
     }
 
-    struct TableElement<Key: store, Value: store> has store {
-        key: Key,
-        value: Value,
-    }
-
-    public fun count<Key: store, Value: store>(table: &Table<Key, Value>): u64 {
-        Vector::length(&table.data)
-    }
-
-    public fun create<Key: store, Value: store>(): Table<Key, Value> {
-        Table {
-            data: Vector::empty(),
+    /// Create a new Table.
+    public fun new<K, V: store>(): Table<K, V> {
+        Table{
+            handle: new_table_handle(),
+            length: 0,
         }
     }
 
-    public fun borrow<Key: store, Value: store>(
-        table: &Table<Key, Value>,
-        key: &Key,
-    ): &Value {
-        let maybe_idx = find(table, key);
-        assert!(Option::is_some(&maybe_idx), Errors::invalid_argument(EKEY_NOT_FOUND));
-        let idx = *Option::borrow(&maybe_idx);
-        &Vector::borrow(&table.data, idx).value
+    /// Destroy a table. The table must be empty to succeed.
+    public fun destroy_empty<K, V>(table: Table<K, V>) {
+        assert!(table.length == 0, Errors::invalid_state(ENOT_EMPTY));
+        destroy_empty_box<K, V, Box<V>>(&table);
+        drop_unchecked_box<K, V, Box<V>>(table)
     }
 
-    public fun borrow_mut<Key: store, Value: store>(
-        table: &mut Table<Key, Value>,
-        key: &Key,
-    ): &mut Value {
-        let maybe_idx = find(table, key);
-        assert!(Option::is_some(&maybe_idx), Errors::invalid_argument(EKEY_NOT_FOUND));
-        let idx = *Option::borrow(&maybe_idx);
-        &mut Vector::borrow_mut(&mut table.data, idx).value
+    /// Add a new entry to the table. Aborts if an entry for this
+    /// key already exists. The entry itself is not stored in the
+    /// table, and cannot be discovered from it.
+    public fun add<K, V>(table: &mut Table<K, V>, key: &K, val: V) {
+        add_box<K, V, Box<V>>(table, key, Box{val});
```

Comment:
What's the behavior if the key already exists?

Reply:
It's an error. https://github.com/move-language/move/blob/be45aa0cc24f8b7106979bdf3313c70eb8d16a12/language/extensions/move-table-extension/src/lib.rs#L222-L226